jax/numpy/lax_numpy.py

```diff
@@ -1021,6 +1022,52 @@ def diff(a, n=1, axis=-1,):
   return a
 
 
+@_wraps(onp.ediff1d)
+def ediff1d(ary, to_end=None, to_begin=None):
+  # convert into 1d array
+  ary = ravel(asarray(ary))
+
+  # default case
+  if to_begin is None and to_end is None:
+    return lax.sub(ary[1:], ary[:-1])
+
+  # enforce propagation of the dtype of input ary to returned array
+  dtype_req = ary.dtype
+
+  if to_begin is None:
+    l_begin = 0
+  else:
+    # check if to_begin can be cast to ary dtype
+    if not can_cast(asarray(to_begin), dtype_req):
+      raise ValueError("cannot convert 'to_begin' to array with dtype "
+              "'%r' as required for input array operand" % dtype_req)
+    # convert to_begin to flat array
+    to_begin = ravel(asarray(to_begin, dtype=dtype_req))
+    l_begin = len(to_begin)
+  
+  if to_end is None:
+    l_end = 0
+  else:
+    # check if to_end can be cast to ary dtype
+    if not can_cast(asarray(to_end), dtype_req):
+      raise ValueError("cannot convert 'to_end' to array with dtype "
+              "'%r' as required for input array operand" % dtype_req)
+    # convert to_end to flat array
+    to_end = ravel(asarray(to_end, dtype=dtype_req))
+    l_end = len(to_end)
+  
+  # calculate difference and copy to_begin and to_end
+  l_diff = _max(len(ary) - 1, 0)
+  result = zeros(l_diff + l_begin + l_end, dtype=ary.dtype)
+  if l_begin > 0:
```

Comment:
Couldn't we write this much more succinctly using `concatenate`?

Reply:
The implementation follows the same strucutre as the numpy implementation.
Nvertheless, I can change into a more compact version!

Reply:
I think in the JAX case `concatenate` is probably preferable, because it's shorter to write and better in terms of, e.g., autodiff.

Reply:
Okay, it is in the last changes incorporated.

Reply:
Just a quick note: I ended up simplifying this in the course of another PR to:

```
_EDIFF1D_DOC = """\
Unlike NumPy's implementation of ediff1d, :py:func:`jax.numpy.ediff1d` will not
issue an error if casting ``to_end`` or ``to_begin`` to the type of ``ary``
loses precision.
"""

@_wraps(onp.ediff1d, lax_description=_EDIFF1D_DOC)
def ediff1d(ary, to_end=None, to_begin=None):
  ary = ravel(asarray(ary))
  result = lax.sub(ary[1:], ary[:-1])
  if to_begin is not None:
    result = concatenate((ravel(asarray(to_begin, dtype=ary.dtype)), result))
  if to_end is not None:
    result = concatenate((result, ravel(asarray(to_end, dtype=ary.dtype))))
  return result
```

It's obviously much shorter (which we like very much!) although one slight downside is it no longer checks `can_cast`. In JAX we tend to like a style that minimizes usage of vertical space.

The issue is with the `can_cast` test is that it is too strict and forbids casts that NumPy would allow. I think it's preferable to allow whatever casts `asarray` would allow, but note in the documentation that the JAX implementation issues errors in fewer places.

Reply:
Okay, perfect!